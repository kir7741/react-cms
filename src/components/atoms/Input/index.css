.input-field {
	display: grid;
	grid-auto-flow: column;
	grid-template-columns: auto 1fr;
	margin-bottom: 10px;
	padding-bottom: 1.25em;
	font-size: 16px;

	&.full {
		grid-template-columns: 1fr;
	}
}

.label {
	display: grid;
	margin-right: 15px;
	align-items: center;
}

.input-wrapper {
	position: relative;

	& span {
		position: absolute;
		bottom: -1.3em;
		left: 0;
		font-size: 14px;
		color: var(--error-red);
		animation-name: slide-in;
		animation-duration: 0.5s;
	}
}

.input {
	padding: 10px;
	border: 1px solid var(--primary-gray);
	border-radius: 6px;
	width: 100%;

	&.error {
		border-color: var(--error-red);
	}

	&:focus {
		outline: none;
	}
}

@keyframes slide-in {
	0% {
		transform: translateY(-100%);
		opacity: 0%;
	}

	30% {
		opacity: 0%;
	}

	100% {
		transform: translateY(0%);
		opacity: 100%;
	}
}
