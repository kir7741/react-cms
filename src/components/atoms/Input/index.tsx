import React, { InputHTMLAttributes, useEffect, useRef } from 'react';
import classnames from 'classnames';
import styles from './index.css';

export interface InputProperty extends InputHTMLAttributes<HTMLInputElement> {

	/**
	 *  class 名稱
	 *
	 * @type {string}
	 * @memberof InputProperty
	 */
	className?: string;

	/**
	 * 輸入框類別
	 *
	 * @type {('text' | 'number' | 'password')}
	 * @memberof InputProperty
	 */
	type?: 'text' | 'number' | 'password';

	/**
	 * label 文字
	 *
	 * @type {string}
	 * @memberof InputProperty
	 */
	labelText?: string;

	/**
	 * 錯誤訊息
	 *
	 * @type {string}
	 * @memberof InputProperty
	 */
	errorMsg?: string;

	/**
	 * 是否在 blur 時觸發檢核
	 *
	 * @type {boolean}
	 * @memberof InputProperty
	 */
	validOnBlur?: boolean;

	/**
	 * 更新表單檢核狀態
	 *
	 * @memberof InputProperty
	 */
	updateCtrlValidity?: () => void,

	/**
	 * change 時觸發的函式
	 *
	 * @memberof InputProperty
	 */
	onChangeValue: (val: string | number) => void

	/**
	 * blur 時觸發的函式
	 *
	 * @memberof InputProperty
	 */
	blur?: () => void

}

/**
 * 輸入框元件
 *
 * @param param0
 * @returns
 */
const Input: React.FC<InputProperty> = ({
	className,
	type = 'text',
	placeholder = '',
	value = '',
	labelText = '',
	errorMsg = '',
	validOnBlur = false,
	disabled,
	updateCtrlValidity = () => {},
	onChangeValue,
	blur = () => {}
}) => {

	const isInit = useRef(true);

	const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		e.preventDefault();
		const val = e.target.value;
		onChangeValue(val);
	};

	useEffect(() => {

		if (isInit.current) {
			isInit.current = false;
			return;
		}

		if (!validOnBlur) {
			updateCtrlValidity();
		}

	}, [value]);

	return (
		<div className={classnames(styles.inputField, className, {
			[styles.full]: !labelText
		})}>
			{labelText && (
				<div className={classnames(styles.label)}>
					{labelText}
				</div>
			)}
			<div className={classnames(styles.inputWrapper)}>
				<input
					type={type}
					disabled={disabled}
					placeholder={placeholder}
					className={classnames(styles.input, errorMsg && styles.error)}
					value={value}
					onBlur={() => {
						if (validOnBlur) {
							updateCtrlValidity();
						}
						blur();
					}}
					onChange={e => handleInputChange(e)}
				/>
				{errorMsg && <span>{errorMsg}</span>}
			</div>
		</div>
	);

};

export default Input;
