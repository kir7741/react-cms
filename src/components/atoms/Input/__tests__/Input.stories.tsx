import type { ComponentStoryObj, ComponentMeta } from '@storybook/react';
import Input from 'components/atoms/Input';

export default {
	title: 'atoms/Input',
	component: Input,
} as ComponentMeta<typeof Input>;

export const Interactive: ComponentStoryObj<typeof Input> = {};
