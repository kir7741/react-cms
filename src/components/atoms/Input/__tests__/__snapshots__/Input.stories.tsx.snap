// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="inputField full"
  >
    <div
      class="inputWrapper"
    >
      <input
        class="input"
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
